spot-vaults/contracts/BillBroker.sol

```diff
@@ -0,0 +1,753 @@
+        if (perpAmtIn <= 0 || usdAmtOut <= 0) {
+            revert UnacceptableSwap();
+        }
+        if (usdAmtOut < usdAmtMin) {
+            revert SlippageTooHigh();
+        }
+
+        // settle fees
+        emit FeeUSD(lpFeeAmt);
+        if (protocolFeeAmt > 0) {
+            usd.safeTransfer(owner(), protocolFeeAmt);
+            emit ProtocolFeeUSD(protocolFeeAmt);
+        }
+
+        // transfer usd out
+        usd.safeTransfer(_msgSender(), usdAmtOut);
+
+        return usdAmtOut;
+    }
+
+    //-----------------------------------------------------------------------------
+    // Public methods
+
+    /// @return s The reserve usd and perp token balances and prices.
+    function reserveState() public returns (ReserveState memory s) {
+        return
+            ReserveState({
+                usdReserve: usdReserve(),
+                perpReserve: perpReserve(),
+                usdPrice: usdPrice(),
+                perpPrice: perpPrice()
+            });
+    }
+
+    /// @dev Reverts if the pricing strategy returns an invalid price.
+    /// @return The price of usd tokens from the pricing strategy.
+    function usdPrice() public returns (uint256) {
+        (uint256 p, bool v) = pricingStrategy.usdPrice();
+        if (!v) {
+            revert UnreliablePrice();
+        }
+        return p;
+    }
+
+    /// @dev Reverts if the pricing strategy returns an invalid price.
+    /// @return The price of perp tokens from the pricing strategy.
+    function perpPrice() public returns (uint256) {
+        (uint256 p, bool v) = pricingStrategy.perpPrice();
+        if (!v) {
+            revert UnreliablePrice();
+        }
+        return p;
+    }
+
+    //-----------------------------------------------------------------------------
+    // Public view methods
+
+    /// @notice Computes the amount of LP tokens minted,
+    ///         when the given number of usd and perp tokens are deposited.
+    /// @param usdAmtMax The maximum available usd tokens.
+    /// @param perpAmtMax The maximum available perp tokens.
+    /// @return mintAmt The amount of LP tokens minted.
+    /// @return usdAmtIn The usd tokens to be deposited.
+    /// @return perpAmtIn The perp tokens to be deposited.
+    function computeMintAmt(
+        uint256 usdAmtMax,
+        uint256 perpAmtMax
+    ) public view returns (uint256 mintAmt, uint256 usdAmtIn, uint256 perpAmtIn) {
+        if (usdAmtMax <= 0 || perpAmtMax <= 0) {
+            return (0, 0, 0);
+        }
+
+        uint256 totalSupply_ = totalSupply();
+        // During the initial deposit we deposit the entire available amounts.
+        // The onus is on the depositor to ensure that the value of USD tokens and
+        // perp tokens on first deposit are equivalent.
+        if (totalSupply_ <= 0) {
+            usdAmtIn = usdAmtMax;
+            perpAmtIn = perpAmtMax;
+            mintAmt = (ONE.mulDiv(usdAmtIn, usdUnitAmt) +
```

Comment:
1) We dont really need to worry about decimals here, but `mintAmt = (ONE.mulDiv(usdAmtIn, usdUnitAmt) + ONE.mulDiv(perpAmtIn, perpUnitAmt))` gives a clean human readable initial denomination for LP tokens. For example, if you lock 100k USD worth of perp and usd, you get 100k LP tokens. (As the pool grows the exchange rate will change anyway and the denomination will no longer make intuitive sense) 

2) I don't think `totalSupply_ < 0` is very expensive, so I'm leaning toward having it in there rather than locking some LP tokens.

Reply:
Yeah I don't think we can get around having a supply <=0 check. We either multiply by 1000 or burn some.

What to do on first deposit, though? I think I'd rather multiply by 1000 than burn some and have 1:1 note:collateral ratio. Even if we control the first deposit on deployment, it could go back to zero again if everyone redeems